proxy/proxy/client.go

```diff
@@ -444,6 +445,32 @@ func (c *Client) selectDialer() func(context.Context, string, string) (net.Conn,
 	}
 }
 
+func (c *Client) invalidateCfg(cfg *tls.Config, instance string) {
+	refreshCfgBuffer := c.RefreshCfgBuffer
+	if refreshCfgBuffer == 0 {
+		refreshCfgBuffer = DefaultRefreshCfgBuffer
+	}
+
+	c.cacheL.RLock()
+	e := c.cfgCache[instance]
+	c.cacheL.RUnlock()
+	if e.cfg != cfg || needsRefresh(e, refreshCfgBuffer) {
+		return
+	}
+	c.cacheL.Lock()
+	e = c.cfgCache[instance]
+	if e.cfg != cfg || needsRefresh(e, refreshCfgBuffer) {
+		c.cacheL.Unlock()
+		return
+	}
+	select {
+	case <-e.done:
+		delete(c.cfgCache, instance)
+	default:
+	}
```

Comment:
Following the logic in my comment above RE: `needsRefresh`, I think that this might also be incorrect:

Specifically at this point, the following assertions are true:
1. The current copy of `cfg` is the same one that lead to a handshake error, so it hasn't been invalidated yet
2. The current copy of `cfg` is valid - aka the next connect won't block in `cachedCfg` on it

I think the desired behavior here is just invalidate the config, but preserve any existing refresh. We also want to preserve `lastUpdated`, so that the refresh throttle isn't bypassed. So I think what we want instead of this select might be:
```go
c.cfgCache[instance] = cacheEntry{
    done: e.done,
    lastUpdated: e.lastUpdated
}
```

I think this can replace the `select` altogether, because if `e.done` is complete, the next connect triggers a refresh ([because it's no longer valid, so `needsRefresh` == true](https://github.com/GoogleCloudPlatform/cloudsql-proxy/blob/60cde814b784f97482d0ff7f8221e725a08471be/proxy/proxy/client.go#L312)). If `e.done` isn't complete, it's also not valid, so will still block as well.

@enocom does that sound right?

Reply:
Yes, I agree. This seems like a cleaner approach.